dwave/optimization/src/nodes/manipulation.cpp

```diff
@@ -18,6 +18,150 @@
+
+        // Arrays must have the same number of dimensions
+        if ((*std::prev(it))->ndim() != (*it)->ndim()) {
+            throw std::invalid_argument(
+                    "all the input arrays must have the same number of dimensions," +
+                    std::string(" but the array at index ") +
+                    std::to_string(std::distance(array_ptrs.begin(), std::prev(it))) +
+                    " has " + std::to_string((*std::prev(it))->ndim()) +
+                    " dimension(s) and the array at index " +
+                    std::to_string(std::distance(array_ptrs.begin(), it)) +
+                    " has " +
+                    std::to_string((*it)->ndim()) +
+                    " dimension(s)");
+        }
+
+        // Array shapes must be the same except for on the concatenation axis
+        for (ssize_t i = 0, stop = (*it)->ndim(); i < stop; ++i) {
+            if (i != axis) {
+                if ( (*std::prev(it))->shape()[i] != (*it)->shape()[i] ) {
+                    throw std::invalid_argument(
+                            "all the input array dimensions except for the concatenation" +
+                            std::string(" axis must match exactly, but along dimension ") +
+                            std::to_string(i) + ", the array at index " +
+                            std::to_string(std::distance(array_ptrs.begin(), std::prev(it))) +
+                            " has size " +
+                            std::to_string((*std::prev(it))->shape()[i]) +
+                            " and the array at index " +
+                            std::to_string(std::distance(array_ptrs.begin(), it)) +
+                            " has size " +
+                            std::to_string((*it)->shape()[i]));
+                }
+            }
+        }
+    }
+
+    // Axis must be in range 0..ndim-1
+    // We can do this check on the first input array since we at
+    // this point know they all have the same number of dimensions
+    if (!(0 <= axis && axis < array_ptrs.front()->ndim())) {
+        throw std::invalid_argument(
+                "axis " +
+                std::to_string(axis) +
+                std::string(" is out of bounds for array of dimension ") +
+                std::to_string(array_ptrs.front()->ndim()));
+    }
+
+
+    // The shape of the input arrays, which will be the
+    // same except for possibly on the concatenation axis
+    std::span<const ssize_t> shape0 = array_ptrs.front()->shape();
+    std::vector<ssize_t> shape(shape0.begin(), shape0.end());
+
+    // On the concatenation axis we sum the axis dimension sizes
+    for (auto it = std::next(array_ptrs.begin()), stop = array_ptrs.end(); it != stop; ++it) {
+       shape[axis] = shape[axis] + (*it)->shape()[axis];
+    }
+
+    return shape;
+}
+
+ConcatenateNode::ConcatenateNode(std::span<ArrayNode*> array_ptrs, const ssize_t axis)
```

Comment:
We should add an explicit error somewhere that all arrays are not dynamic. That is `array_ptr->size() >= 0`.

Reply:
Not sure I follow, does `array_ptr->size() < 0` mean the array is dynamic?

Reply:
Yes. `array_ptr->size()` will be `-1` if the size of the array is dependent on the state. `array_ptr->size(state)` will always return a non-negative number.

Reply:
Fixed! Using `(*ptr)->dynamic()`  [e7213d6](https://github.com/dwavesystems/dwave-optimization/pull/153/commits/e7213d6ab70b55cab35df1273a65babd43e5f069)